sympy/physics/units/dimensions.py

```diff
@@ -572,3 +572,126 @@ def is_consistent(self):
         # dimensions
         # in vector language: the set of vectors do not form a basis
         return self.inv_can_transf_matrix.is_square
+
+    def buckingham_pi_theorem(self, list_of_derived_quantities):
+        """
+        list_of_derived_quantities is the input of derived quantities in the form of a list of tuples (quantity_name, dims) where quantity_name
+        is the name of the physical quantity and dims is an instance of Dimension class which possesses the dimensions of the quantity. This function
+        is designed to take in a set of physical variables (all essentially dimensionful) and returns one set of all possible exponents,
+        as guided by the Buckingham's pi theorem (raise the list of quantities to the list of exponents to get the dimensionless numbers).
+
+        The next function verify_dimensionless_numbers() is used to check whether the dimensions computed after their product with the
+        exponent matrix (as guided by the Buckingham's pi theorem). Finally the function get_dimensionless_numbers() is used to
+        produce the requisite result in the form of a list of instances of Dimension class signifying the actual dimensionless quantities.
+
+        For example, consider the scenario in the case of a fluid dynamics problem governed by Navier-Stokes equations for
+        an incompressible Newtonian fluid in an SI system ....
+
+            >>> from sympy.physics.units import length, mass, time
+            >>> from sympy.physics.units.systems.si import dimsys_SI
+            >>> M = mass
+            >>> L = length
+            >>> T = time
+            >>> Lc = L # Characteristic length scale
+            >>> Uc = L/T # Characteristic velocity scale
+            >>> rho_c = M/L**3 # Characteristic density scale
+            >>> mu = M/L/T # Dynamic viscosity
+            >>> p_c = M/L/T**2 # Characteristic pressure scale
+            >>> F_bc = M*L/T**2 # Characteristic body forces scale
+            >>> list_of_quantities = [('Lc',Lc), ('Uc',Uc), ('rho',rho_c), ('mu',mu), ('p_c',p_c), ('Fbc',F_bc)]
+            >>> dimsys_SI.verify_dimensionless_numbers(list_of_quantities)
+            [Dimension(1), Dimension(1), Dimension(1)]
+            >>> set_of_dimless_nums = dimsys_SI.verify_dimensionless_numbers(list_of_quantities)
+            >>> flag = True
+            >>> for dimension in set_of_dimless_nums:
+            ...     flag = flag and dimsys_SI.get_dimensional_dependencies(dimension) == {}
+            >>> flag
```

Comment:
This code is definitely hard to understand... This should not go into the documentation. Can you rather think of some easy examples of applications that a SymPy user may need that can be explained with some theory in the comments?